### PR TITLE
Integrate appdynamics

### PR DESCRIPTION
**Pull request by MALPI**
This is an attempt to integrate appdynamics agent. Please test this first!

### CODE REVIEW

**Comment by lmineiro** [comment]
Typo in the filename `utils/appdynmaics-agent`

File: create_cluster.py [hunk 1]
```diff
@@ -576,6 +578,9 @@ def cli(cluster_name: str, regions: list, cluster_size: int, instance_type: str,
     if not re.match(cluster_name_re, cluster_name):
         raise click.UsageError('Cluster name must only contain lowercase latin letters, digits and dashes (it also must start with a letter and cannot end with a dash), in other words it must be matched by the following regular expression: {}'.format(cluster_name_re))
 
+    if not application_name:
+        raise.click.UsageError('Please specify application name which shell be used.')
```
Comment (lmineiro) [inline create_cluster.py]: Typo "shell". Simplify to 'Please specify the application name to be used'

Comment (lmineiro) [inline create_cluster.py]: Also, I believe the cluster-name is already used as a unique KIO application_id. Can't this be the the AD application name? It's already creepy the amount of parameters we have

Comment (MALPI) [inline create_cluster.py]: Use case here is actually, that for example our team is running all apps with one Application ID.

**Comment by lmineiro** [comment]
My opinion is that we shouldn't support the appdynamics agent. Can you describe which particular features of AD you're looking for?

C* is a highly performant system where the agent overhead may not be negligible. I understand how the the idea of "free metrics" can be inviting but I personally don't believe it will send anyone patching C* code.

That being said, if this still gets merged, it doesn't make sense to have both Jolokia and AD Agent in the same cluster. Jolokia is there mostly to enable easy cluster metrics extraction.

**Comment by MALPI** [comment]
@lmineiro I totally agree. Unfortunately the issue is, that we have to use logging of appdynamics. Therefore the appdynamics agent is needed.

File: create_cluster.py (hunk 1 above)
Comment (lmineiro) [inline create_cluster.py]: Can you make this optional? I don't think everyone in the world will want this.

Comment (MALPI) [inline create_cluster.py]: Done

**Comment by hjacobs** [comment]
@MALPI logging is supported by Taupage and the AppDynamics machine agent --- no need to integrate AD into the Docker image itself.

File: create_cluster.py [hunk 2]
```diff
@@ -563,11 +564,12 @@ def print_failure_message():
 @click.option('--docker-image', help='Docker image to use (default: use latest planb-cassandra)')
 @click.option('--sns-topic', help='SNS topic name to send Auto-Recovery notifications to')
 @click.option('--sns-email', help='Email address to subscribe to Auto-Recovery SNS topic')
+@click.option('--application-name', help='Please specify the application name to be used')
```
Comment (hjacobs) [inline create_cluster.py]: maybe rename to "appdynamics-application"? "application-name" alone is confusing (conflicts with Kio application ID)

Comment (MALPI) [inline create_cluster.py]: Done.

**Comment by MALPI** [comment]
Should I also drop `scalyr_key`?

**Comment by lmineiro** [comment]
Why? As a feature, someone may be using Taupage that supports it out there.

File: create_cluster.py [hunk 3]
```diff
@@ -563,11 +564,12 @@ def print_failure_message():
 @click.option('--docker-image', help='Docker image to use (default: use latest planb-cassandra)')
 @click.option('--sns-topic', help='SNS topic name to send Auto-Recovery notifications to')
 @click.option('--sns-email', help='Email address to subscribe to Auto-Recovery SNS topic')
+@click.option('--appdynamics-application', help='Please specify the appdynamics application name to be used')
 @click.argument('regions', nargs=-1)
 def cli(cluster_name: str, regions: list, cluster_size: int, instance_type: str,
         volume_type: str, volume_size: int, volume_iops: int,
         no_termination_protection: bool, internal: bool, hosted_zone: str, scalyr_key: str,
-        docker_image: str, sns_topic: str, sns_email: str):
+        docker_image: str, sns_topic: str, sns_email: str, application_name: str):
```
Comment (hjacobs) [inline create_cluster.py]: this will not match, you need to rename to "appdynamics_application" too

**Comment by MALPI** [comment]
Ok.

File: create_cluster.py (hunk 3 above)
Comment (MALPI) [inline create_cluster.py]: fixed

**Comment by hjacobs** [comment]
:+1:

**Comment by lmineiro** [comment]
👍

**Comment by hjacobs** [comment]
arrrrrrrrrrrr

```
 $ ./create_cluster.py --help
  File "./create_cluster.py", line 361
    'appdynamics_application': if options['appdynamics_application'] else options['cluster_name'],
                                ^
SyntaxError: invalid syntax

**Comment by a1exsh** [comment]
flake8 to the rescue? ;-)

**Comment by hjacobs** [comment]
@a1exsh yes + automated tests :smirk:

**Pull request merged**